src/peer_list.cpp

```diff
@@ -79,7 +79,64 @@ namespace {
 		address const& m_addr;
 		std::uint16_t m_port;
 	};
-}
+
+	// this returns true if lhs is a better erase candidate than rhs
+	bool compare_peer_erase(torrent_peer const& lhs, torrent_peer const& rhs)
+	{
+		TORRENT_ASSERT(lhs.connection == nullptr);
+		TORRENT_ASSERT(rhs.connection == nullptr);
+
+		// primarily, prefer getting rid of peers we've already tried and failed
+		if (lhs.failcount != rhs.failcount)
+			return lhs.failcount > rhs.failcount;
+
+		bool const lhs_resume_data_source = lhs.peer_source() == peer_info::resume_data;
+		bool const rhs_resume_data_source = rhs.peer_source() == peer_info::resume_data;
+
+		// prefer to drop peers whose only source is resume data
+		if (lhs_resume_data_source != rhs_resume_data_source)
+			return int(lhs_resume_data_source) > int(rhs_resume_data_source);
+
+		if (lhs.connectable != rhs.connectable)
+			return int(lhs.connectable) < int(rhs.connectable);
+
+		return lhs.trust_points < rhs.trust_points;
+	}
+
+	// this returns true if lhs is a better connect candidate than rhs
+	bool compare_peer(torrent_peer const* lhs, torrent_peer const* rhs
+		, external_ip const& external, int const external_port, bool const finished)
+	{
+		// prefer peers with lower failcount
+		if (lhs->failcount != rhs->failcount)
+			return lhs->failcount < rhs->failcount;
+
+		// Local peers should always be tried first
+		bool const lhs_local = is_local(lhs->address());
+		bool const rhs_local = is_local(rhs->address());
+		if (lhs_local != rhs_local) return int(lhs_local) > int(rhs_local);
+
+		if (lhs->last_connected != rhs->last_connected)
+			return lhs->last_connected < rhs->last_connected;
+
+		if (lhs->maybe_seed != rhs->maybe_seed)
+		{
+			// if we're finished, de-prioritze peers we think may be seeds
+			// otherwise, prioritize peers we think may be seeds
+			if (finished) return rhs->maybe_seed;
+			else return lhs->maybe_seed;
```

Comment:
given that the spec says that this may also include upload-only peers, this inverse prioritization should probably not be made